identity-iota/src/tangle/client.rs

```diff
@@ -97,6 +105,27 @@ impl Client {
       .map(|message| Receipt::new(self.network.clone(), message))
   }
 
+  /// Publishes arbitrary JSON data to the specified index on the Tangle.
+  /// Retries (promotes or reattaches) the message until it’s included (referenced by a milestone).
+  /// Default interval is 5 seconds and max attempts is 40.
+  pub async fn publish_json_with_retry<T: ToJson>(
+    &self,
+    index: &str,
+    data: &T,
+    interval: Option<u64>,
+    max_attempts: Option<u64>,
+  ) -> Result<Receipt> {
+    let receipt: Receipt = self.publish_json(index, data).await?;
+    let reattached_messages: Vec<(MessageId, Message)> = self
+      .client
+      .retry_until_included(receipt.message_id(), interval, max_attempts)
+      .await?;
+    match reattached_messages.into_iter().next() {
+      Some((_, message)) => Ok(Receipt::new(self.network.clone(), message)),
+      None => Ok(receipt),
+    }
```

Comment:
Do we not have to handle the `Error::TangleInclusionError` that can be returned from `retry_until_included`? I'm not entirely sure if that method can still return this error, even if `publish_json` successfully included our message.

If that error is returned, we may have to do another call to check the `LedgerInclusionStateDto` of the message, as it is done [here](https://github.com/iotaledger/iota.rs/tree/b9dbaa735963e88a061f44bc5fdce6b0fe8463e9/iota-client/src/client.rs#L1341-L1344), but I haven't looked too deep into the logic of `retry_until_included`.

Also, if it is guaranteed that `retry_until_included` returns the included message on `Ok`, then we technically don't have to handle the case where `reattached_messages.into_iter().next()` returns `None`, although we also shouldn't panic, so mapping it to some error seems like the best we can do.

Reply:
Talked to Thoralf about this.

- It's practically guaranteed that the returned vector is non-empty, but we should handle this anyway and map it to some reasonable error. 
- And the retry function can return a `TangleInclusionError`, but the message might have been included with `publish_json` already, so I think it'd be good to add the additional inclusion check.